### DISCUSSION TITLE
STL reimplementation questions

### DISCUSSION THREAD

**Discussion by danlapid** [jzimmerman/langcc#34] [Q&A] [answered]
Hey @jzimmerman,
I would love to know your considerations behind reimplementing many STL types given you're depending on C++17 and using these STL types elsewhere anyway.
To be clear I mean 
Vec instead of std::vector
rc_ptr instead of std::shared_ptr
StrSlice instead of std::string
Option instead of std::optional
_W+is_*+as_*+match() instead of std::variant, std::holds_alternative, std::get, std::visit

I might be missing some but I assume your answer would be general anyway.
Seems to me that switching over to those types would simplify the codebase (or at least shorten it) significantly.
Regardless, amazing job, I enjoy using it a lot.

**Comment by jzimmerman** [comment, marked as answer]
In many cases, this is done for performance reasons. When I was profiling the execution of generated parsers, I found that a large amount of time was spent in STL primitives, and I was able to cut that down significantly by implementing custom versions. This is particularly apparent for types like StrSlice (which pass around the indices into a shared string, instead of making copies of the substrings), but performance issues also arose in less obvious instances, like std::vector iterators.

In other cases, it's done for semantic reasons. For instance, I would much rather just pass around reference-counted pointers to large objects than mess with move semantics (which can cause subtle undefined behavior if you get it wrong). Similarly, for std::variant, I wanted a _named_ sum type, i.e., one in which different cases could theoretically have the same underlying type. Doing this with std::variant would require a lot of boilerplate.

**Comment by danlapid** [comment]
STL is known to have subpar performance on debug builds to be sure but should be quite optimal on release builds (mainly about inlining not happening)
I noticed the build script always builds debug
I wonder if the same results would be replicated in release builds with O2 optimization 
Regarding references to shared string you could consider ranges views

The problem you referred to regarding move semantics on large sized objects should be solved just the same with shared_pointer where move/copy is almost identical

I understand the problem regarding std variant though, seems to indeed not serve the purpose here

**Reply by jzimmerman** [reply]
The build script builds with debugging information enabled (`-g`, etc.) but still builds a release build (`-O3`), so that should not be the issue.

> The problem you referred to regarding move semantics on large sized objects should be solved just the same with shared_pointer where move/copy is almost identical

It's true that one could pass, e.g., `shared_ptr<vector<T>>` to solve this problem, but that would still leave the underlying performance issues with `shared_ptr` and `std::vector`.
